src/cellarr/dataloader.py

```diff
@@ -1,15 +1,374 @@
+import numpy as np
+import os
+import pytorch_lightning as pl
+from scipy.sparse import coo_matrix, csr_matrix
 import tiledb
+import torch
+from torch.utils.data import Dataset, DataLoader, WeightedRandomSampler
+from typing import List, Optional
+
+from cellarr.queryutils_tiledb_frame import subset_frame
 
 __author__ = "Tony Kuo"
 __copyright__ = "Jayaram Kancherla"
 __license__ = "MIT"
 
 # Turn off multithreading to allow multiple pytorch dataloader workers
 config = tiledb.Config()
-# config["sm.compute_concurrency_level"] = 1
-# config["sm.io_concurrency_level"] = 1
-# config["sm.num_async_threads"] = 1
-# config["sm.num_reader_threads"] = 1
-# config["sm.num_tbb_threads"] = 1
-# config["sm.num_writer_threads"] = 1
+config["sm.compute_concurrency_level"] = 1
+config["sm.io_concurrency_level"] = 1
+config["sm.num_async_threads"] = 1
+config["sm.num_reader_threads"] = 1
+config["sm.num_tbb_threads"] = 1
+config["sm.num_writer_threads"] = 1
 config["vfs.num_threads"] = 1
+
+
+class scDataset(Dataset):
+    """A pytorch Dataset that enumerates cells and cell labels in TileDB."""
+
+    def __init__(
+        self,
+        data_df: "pandas.DataFrame",
+        matrix_tdb: "tiledb.libtiledb.SparseArrayImpl",
+        matrix_shape: tuple,
+        gene_indices: List[int],
+        label_column: str,
+        study_column: str,
+        lognorm: bool = True,
+        target_sum: float = 1e4,
+    ):
+        """Constructor.
+
+        Parameters
+        ----------
+        data_df: pandas.DataFrame
+            Pandas dataframe of valid cells.
+        matrix_tdb: tiledb.libtiledb.SparseArrayImpl
+            Counts TileDB.
+        matrix_shape: tuple,
+            Shape of the counts matrix
+        gene_indices: List[int]
+            The index of genes to return.
+        label_column: str, default: "cellTypeName"
+            Label column name.
+        study_column: str, default: "dataset"
+            Study column name.
+        lognorm: bool, default: True
+            Whether to return log normalized expression instead of raw counts.
+        target_sum: float, default: 1e4
+            Target sum for log normalization.
+        """
+
+        self.data_df = data_df
+        self.matrix_tdb = matrix_tdb
+        self.matrix_shape = matrix_shape
+        self.gene_indices = gene_indices
+        self.lognorm = lognorm
+        self.target_sum = target_sum
+        self.label_column = label_column
+        self.study_column = study_column
+
+    def __len__(self):
+        return self.data_df.shape[0]
+
+    def __getitem__(self, idx):
+        # data, label, study
+        cell_idx = self.data_df.index[idx]
+        results = self.matrix_tdb.multi_index[cell_idx, :]
+        counts = coo_matrix(
+            (results["data"], (results["cell_index"], results["gene_index"])),
+            shape=self.matrix_shape,
+        ).tocsr()
+        counts = counts[cell_idx, :]
+        counts = counts[:, self.gene_indices]
```

Comment:
wonder if we can pass this on to https://github.com/BiocPy/cellarr/blob/master/src/cellarr/CellArrDataset.py#L305; may simplify things for you especially with how the shape is manipulated

Reply:
That would be useful I think.  Just accessing 1 cell by index.